File: typescript/tests/goal_progress.test.ts
// goal_progress.test.ts
//
// Cross-language Goal.advanceProgress() verification.
// All test vectors from fixtures/goal_progress.json.
// Formula: new_progress = clamp(progress + delta, 0.0, 1.0)

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import * as fs from 'fs';
import * as path from 'path';
import { Goal, GoalStatus, GoalPriority } from '../src/memory';

// ---------------------------------------------------------------------------
// Fixture types
// ---------------------------------------------------------------------------

interface GoalProgressVector {
  id:               string;
  description:      string;
  initial_progress: number;
  delta:            number;
  expected_progress: number;
}

interface GoalProgressFixture {
  vectors: GoalProgressVector[];
}

// ---------------------------------------------------------------------------
// Load fixture
// ---------------------------------------------------------------------------

const fixturePath = path.join(__dirname, '..', '..', 'fixtures', 'goal_progress.json');
const fixture: GoalProgressFixture = JSON.parse(fs.readFileSync(fixturePath, 'utf-8'));

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function makeGoal(progress: number): Goal {
  const g = new Goal();
  g.id          = 'test-goal-001';
  g.userId      = 'test-user';
  g.title       = 'Test Goal';
  g.description = 'A goal for testing advance progress';
  g.status      = GoalStatus.Active;
  g.priority    = GoalPriority.Normal;
  g.createdUtc  = new Date('2026-01-01T00:00:00Z');
  g.progress    = progress;
  return g;
}

function checkClose(got: number, want: number, epsilon: number, label: string): void {
  assert.ok(
    Math.abs(got - want) <= epsilon,
    `${label}: got ${got}, want ${want} (diff ${Math.abs(got - want)} > ${epsilon})`,
  );
}

const EPSILON = 1e-5;

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

describe('Goal.advanceProgress fixture vectors', () => {
  for (const vec of fixture.vectors) {
    it(`${vec.id} — ${vec.description}`, () => {
      const original = makeGoal(vec.initial_progress);
      const updated  = original.advanceProgress(vec.delta);

      checkClose(updated.progress, vec.expected_progress, EPSILON, 'progress');

      // advanceProgress must return a NEW Goal, not mutate the original.
      assert.equal(
        original.progress,
        vec.initial_progress,
        `advanceProgress must not mutate original goal: expected ${vec.initial_progress}, got ${original.progress}`,
      );
    });
  }
});

describe('Goal.advanceProgress immutability', () => {
  it('returns a new Goal instance', () => {
    const g1 = makeGoal(0.5);
    const g2 = g1.advanceProgress(0.1);
    assert.ok(g1 !== g2, 'advanceProgress must return a different object');
  });

  it('preserves all other fields on the returned Goal', () => {
    const g1   = makeGoal(0.3);
    const g2   = g1.advanceProgress(0.2);
    assert.equal(g2.id,          g1.id);
    assert.equal(g2.userId,      g1.userId);
    assert.equal(g2.title,       g1.title);
    assert.equal(g2.description, g1.description);
    assert.equal(g2.status,      g1.status);
    assert.equal(g2.priority,    g1.priority);
  });
});

describe('Goal.advanceProgress clamping', () => {
  it('result is always >= 0', () => {
    const g = makeGoal(0.0);
    assert.ok(g.advanceProgress(-100).progress >= 0, 'progress should not go below 0');
  });

  it('result is always <= 1', () => {
    const g = makeGoal(1.0);
    assert.ok(g.advanceProgress(100).progress <= 1, 'progress should not exceed 1');
  });
});

describe('Goal default values', () => {
  it('default progress is 0.0', () => {
    const g = new Goal();
    assert.equal(g.progress, 0.0);
  });

  it('default status is Active', () => {
    const g = new Goal();
    assert.equal(g.status, GoalStatus.Active);
  });

  it('default priority is Normal', () => {
    const g = new Goal();
    assert.equal(g.priority, GoalPriority.Normal);
  });
});

File: rust/tests/goal_progress_test.rs
//! goal_progress_test.rs
//!
//! 7 cross-language test vectors for Goal::advance_progress,
//! loaded from fixtures/goal_progress.json.
//! Progress is clamped to [0.0, 1.0].

use circle_ai::memory::{Goal, GoalPriority};
use serde::Deserialize;

const EPSILON: f32 = 1e-6_f32;

// ─────────────────────────────────────────────────────────────────────────────
// Fixture deserialization
// ─────────────────────────────────────────────────────────────────────────────

#[derive(Debug, Deserialize)]
struct ProgressVector {
    id: String,
    initial_progress: f32,
    delta: f32,
    expected_progress: f32,
}

#[derive(Debug, Deserialize)]
struct Fixture {
    vectors: Vec<ProgressVector>,
}

fn load_fixture() -> Fixture {
    let fixtures_dir = std::path::Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .unwrap()
        .join("fixtures");
    let path = fixtures_dir.join("goal_progress.json");
    let text = std::fs::read_to_string(&path)
        .unwrap_or_else(|e| panic!("Failed to read {:?}: {}", path, e));
    serde_json::from_str(&text).expect("Failed to parse goal_progress.json")
}

fn make_goal(progress: f32) -> Goal {
    let mut g = Goal::new("g1", "u1", "Test goal", "A test goal description", GoalPriority::Normal);
    g.progress = progress;
    g
}

// ─────────────────────────────────────────────────────────────────────────────
// Fixture-driven test
// ─────────────────────────────────────────────────────────────────────────────

#[test]
fn test_all_progress_vectors() {
    let fixture = load_fixture();
    assert_eq!(
        fixture.vectors.len(),
        7,
        "Expected 7 progress vectors, found {}",
        fixture.vectors.len()
    );

    for v in &fixture.vectors {
        let goal = make_goal(v.initial_progress);
        let advanced = goal.advance_progress(v.delta);
        let diff = (advanced.progress - v.expected_progress).abs();
        assert!(
            diff <= EPSILON,
            "[{}] progress: got {}, expected {}, diff={}",
            v.id,
            advanced.progress,
            v.expected_progress,
            diff
        );
        // Original goal must be unchanged (advance_progress returns a clone)
        assert_eq!(
            goal.progress, v.initial_progress,
            "[{}] original goal progress mutated",
            v.id
        );
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Individual pinned tests
// ─────────────────────────────────────────────────────────────────────────────

#[test]
fn test_zero_initial_zero_delta() {
    let g = make_goal(0.0);
    assert!((g.advance_progress(0.0).progress - 0.0).abs() <= EPSILON);
}

#[test]
fn test_partial_advance_from_zero() {
    let g = make_goal(0.0);
    let result = g.advance_progress(0.3);
    assert!((result.progress - 0.3).abs() <= EPSILON, "got {}", result.progress);
}

#[test]
fn test_clamp_at_max() {
    let g = make_goal(0.9);
    let result = g.advance_progress(0.5);
    assert!(
        (result.progress - 1.0).abs() <= EPSILON,
        "expected 1.0, got {}",
        result.progress
    );
}

#[test]
fn test_clamp_at_min() {
    let g = make_goal(0.1);
    let result = g.advance_progress(-0.5);
    assert!(
        (result.progress - 0.0).abs() <= EPSILON,
        "expected 0.0, got {}",
        result.progress
    );
}

#[test]
fn test_zero_delta_mid_progress() {
    let g = make_goal(0.5);
    let result = g.advance_progress(0.0);
    assert!((result.progress - 0.5).abs() <= EPSILON, "got {}", result.progress);
}

#[test]
fn test_advance_to_exactly_full() {
    let g = make_goal(0.75);
    let result = g.advance_progress(0.25);
    assert!(
        (result.progress - 1.0).abs() <= EPSILON,
        "expected 1.0, got {}",
        result.progress
    );
}

#[test]
fn test_negative_delta_no_floor_hit() {
    let g = make_goal(0.6);
    let result = g.advance_progress(-0.2);
    assert!(
        (result.progress - 0.4).abs() <= EPSILON,
        "expected 0.4, got {}",
        result.progress
    );
}

#[test]
fn test_advance_progress_does_not_mutate_original() {
    let g = make_goal(0.3);
    let _advanced = g.advance_progress(0.4);
    assert!(
        (g.progress - 0.3).abs() <= EPSILON,
        "original goal should not be mutated, got {}",
        g.progress
    );
}
